File: flutter_app/lib/models/task.dart
class Task {
  int id;           // local SQLite ID (primary key in app)
  int? serverId;    // cloud server ID (null = not synced yet)
  bool synced;      // true = local matches cloud

  String title;
  String body;
  String color;
  String font;
  int fontSize;
  String textColor;
  bool isBold;
  bool isItalic;
  bool isUnderline;
  String textAlign;
  String emojiStamp;
  double posX;
  double posY;
  double rotation;
  bool isDone;
  DateTime? reminderTime;
  String tag;
  final DateTime createdAt;

  Task({
    required this.id,
    this.serverId,
    this.synced = false,
    required this.title,
    this.body = '',
    this.color = '#FFF9C4',
    this.font = 'Caveat',
    this.fontSize = 16,
    this.textColor = '#212121',
    this.isBold = false,
    this.isItalic = false,
    this.isUnderline = false,
    this.textAlign = 'left',
    this.emojiStamp = '',
    this.posX = 50,
    this.posY = 100,
    this.rotation = 0,
    this.isDone = false,
    this.reminderTime,
    this.tag = '',
    required this.createdAt,
  });

  // From cloud API response
  factory Task.fromJson(Map<String, dynamic> j) => Task(
        id:           j['id'],
        serverId:     j['id'],
        synced:       true,
        title:        j['title'],
        body:         j['body'] ?? '',
        color:        j['color'] ?? '#FFF9C4',
        font:         j['font'] ?? 'Caveat',
        fontSize:     j['font_size'] ?? 16,
        textColor:    j['text_color'] ?? '#212121',
        isBold:       j['is_bold'] ?? false,
        isItalic:     j['is_italic'] ?? false,
        isUnderline:  j['is_underline'] ?? false,
        textAlign:    j['text_align'] ?? 'left',
        emojiStamp:   j['emoji_stamp'] ?? '',
        posX:         (j['pos_x'] as num).toDouble(),
        posY:         (j['pos_y'] as num).toDouble(),
        rotation:     (j['rotation'] as num).toDouble(),
        isDone:       j['is_done'] ?? false,
        reminderTime: j['reminder_time'] != null
            ? DateTime.parse(j['reminder_time'])
            : null,
        tag:          j['tag'] ?? '',
        createdAt:    DateTime.parse(j['created_at']),
      );

  // From local SQLite row
  factory Task.fromDb(Map<String, dynamic> row) => Task(
        id:           row['id'] as int,
        serverId:     row['server_id'] as int?,
        synced:       (row['synced'] as int) == 1,
        title:        row['title'] as String,
        body:         row['body'] as String? ?? '',
        color:        row['color'] as String? ?? '#FFF9C4',
        font:         row['font'] as String? ?? 'Caveat',
        fontSize:     row['font_size'] as int? ?? 16,
        textColor:    row['text_color'] as String? ?? '#212121',
        isBold:       (row['is_bold'] as int? ?? 0) == 1,
        isItalic:     (row['is_italic'] as int? ?? 0) == 1,
        isUnderline:  (row['is_underline'] as int? ?? 0) == 1,
        textAlign:    row['text_align'] as String? ?? 'left',
        emojiStamp:   row['emoji_stamp'] as String? ?? '',
        posX:         (row['pos_x'] as num?)?.toDouble() ?? 50,
        posY:         (row['pos_y'] as num?)?.toDouble() ?? 100,
        rotation:     (row['rotation'] as num?)?.toDouble() ?? 0,
        isDone:       (row['is_done'] as int? ?? 0) == 1,
        reminderTime: row['reminder_time'] != null
            ? DateTime.tryParse(row['reminder_time'] as String)
            : null,
        tag:          row['tag'] as String? ?? '',
        createdAt:    DateTime.parse(row['created_at'] as String),
      );

  Map<String, dynamic> toJson() => {
        'title':        title,
        'body':         body,
        'color':        color,
        'font':         font,
        'font_size':    fontSize,
        'text_color':   textColor,
        'is_bold':      isBold,
        'is_italic':    isItalic,
        'is_underline': isUnderline,
        'text_align':   textAlign,
        'emoji_stamp':  emojiStamp,
        'pos_x':        posX,
        'pos_y':        posY,
        'rotation':     rotation,
        'is_done':      isDone,
        'tag':          tag,
      };

  Map<String, dynamic> toDbRow() => {
        if (serverId != null) 'server_id': serverId,
        'title':        title,
        'body':         body,
        'color':        color,
        'font':         font,
        'font_size':    fontSize,
        'text_color':   textColor,
        'is_bold':      isBold ? 1 : 0,
        'is_italic':    isItalic ? 1 : 0,
        'is_underline': isUnderline ? 1 : 0,
        'text_align':   textAlign,
        'emoji_stamp':  emojiStamp,
        'pos_x':        posX,
        'pos_y':        posY,
        'rotation':     rotation,
        'is_done':      isDone ? 1 : 0,
        'reminder_time': reminderTime?.toIso8601String(),
        'tag':          tag,
        'synced':       synced ? 1 : 0,
        'updated_at':   DateTime.now().toIso8601String(),
      };
}

File: backend/models/task.py
from datetime import datetime, timezone
from extensions import db


class Task(db.Model):
    __tablename__ = "tasks"

    id            = db.Column(db.Integer, primary_key=True)
    user_id       = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=True)
    title         = db.Column(db.String(200), nullable=False)
    body          = db.Column(db.Text, default="")
    color         = db.Column(db.String(20), default="#FFF9C4")   # hex pastel
    font          = db.Column(db.String(60), default="Caveat")
    font_size     = db.Column(db.Integer, default=16)
    text_color    = db.Column(db.String(20), default="#212121")
    is_bold       = db.Column(db.Boolean, default=False)
    is_italic     = db.Column(db.Boolean, default=False)
    is_underline  = db.Column(db.Boolean, default=False)
    text_align    = db.Column(db.String(10), default="left")
    emoji_stamp   = db.Column(db.String(10), default="")          # single emoji
    pos_x         = db.Column(db.Float, default=50.0)
    pos_y         = db.Column(db.Float, default=100.0)
    rotation      = db.Column(db.Float, default=0.0)              # degrees
    is_done       = db.Column(db.Boolean, default=False)
    reminder_time = db.Column(db.DateTime, nullable=True)
    tag           = db.Column(db.String(60), default="")
    created_at    = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at    = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc),
                              onupdate=lambda: datetime.now(timezone.utc))

    def to_dict(self):
        return {
            "id":           self.id,
            "title":        self.title,
            "body":         self.body,
            "color":        self.color,
            "font":         self.font,
            "font_size":    self.font_size,
            "text_color":   self.text_color,
            "is_bold":      self.is_bold,
            "is_italic":    self.is_italic,
            "is_underline": self.is_underline,
            "text_align":   self.text_align,
            "emoji_stamp":  self.emoji_stamp,
            "pos_x":        self.pos_x,
            "pos_y":        self.pos_y,
            "rotation":     self.rotation,
            "is_done":      self.is_done,
            "reminder_time": self.reminder_time.isoformat() if self.reminder_time else None,
            "tag":          self.tag,
            "created_at":   self.created_at.isoformat(),
            "updated_at":   self.updated_at.isoformat(),
        }
